### ISSUE TITLE
[bst-1] "Error instantiating artifact cache: Size '' parsed from '/home/fedora/.cache/buildstream/artifacts/cache_size' was not an integer"

### ISSUE THREAD

**Issue by devcurmudgeon** [apache/buildstream#1540]
I had to shut down server while it was in the middle of running a bst (1.6.3) command. On restart, bst fails with

"Error instantiating artifact cache: Size '' parsed from '/home/fedora/.cache/buildstream/artifacts/cache_size' was not an integer"

I daresay the cache is somehow corrupted, but it should not be possible to get into a state where bst can't recover

**Comment by devcurmudgeon** [comment]
on deleting the file '/home/fedora/.cache/buildstream/artifacts/cache_size' restarting the build failed in a way which suggests both that the cache is corrupted, and bst does not know what to do about it. Neither of these situations should be possible imo (i.e cache objects should be guaranteed correct if they exist, and bst should be able to reconstruct any cache artifact in the event of a problem)

```
[--:--:--][7dd2aff9][build:components/nasm.bst           ] STATUS  Running integration command

    set -e
    
    [ -f "/etc/xml/catalog" ] || xmlcatalog --noout --create "/etc/xml/catalog"
    add_main() {
      xmlcatalog --noout --add "$1" "$2" "file:///usr/share/xml/docbook/xml/xsl-stylesheets" "/etc/xml/catalog"
    }
    
    add_main rewriteSystem "http://docbook.sourceforge.net/release/xsl/1.79.2"
    add_main rewriteURI "http://docbook.souceforge.net/release/xsl/1.79.2"
    add_main rewriteSystem "http://docbook.sourceforge.net/release/xsl/current"
    add_main rewriteURI "http://docbook.sourceforge.net/release/xsl/current"

[00:00:00][7dd2aff9][build:components/nasm.bst           ] FAILURE Integrating sandbox
[00:00:03][7dd2aff9][build:components/nasm.bst           ] FAILURE Could not find file at /home/fedora/.cache/buildstream/artifacts/extract/freedesktop-sdk/components-asciidoc/35c899591f013697752eea1aa856cac2aa688073f4f129fc7a7a56f44aa3790d/meta/public.yaml

    Sandbox directory: /home/fedora/.cache/buildstream/build/components-nasm-3ajpse39

    Printing the last 20 lines from log file:
    /home/fedora/.cache/buildstream/logs/freedesktop-sdk/components-nasm/7dd2aff9-build.1915.log
    ======================================================================
    
        add_main delegatePublic "-//OASIS//ENTITIES DocBook XML"
        add_main delegatePublic "-//OASIS//DTD DocBook XML"
        add_main delegateSystem "http://www.oasis-open.org/docbook/"
        add_main delegateSystem "http://docbook.org/xml/"
    [--:--:--] STATUS  components/docbook-xsl.bst: Running integration command

        set -e
    
        [ -f "/etc/xml/catalog" ] || xmlcatalog --noout --create "/etc/xml/catalog"
        add_main() {
          xmlcatalog --noout --add "$1" "$2" "file:///usr/share/xml/docbook/xml/xsl-stylesheets" "/etc/xml/catalog"
        }
    
        add_main rewriteSystem "http://docbook.sourceforge.net/release/xsl/1.79.2"
        add_main rewriteURI "http://docbook.souceforge.net/release/xsl/1.79.2"
        add_main rewriteSystem "http://docbook.sourceforge.net/release/xsl/current"
        add_main rewriteURI "http://docbook.sourceforge.net/release/xsl/current"
    [00:00:00] FAILURE components/nasm.bst: Integrating sandbox
    [00:00:03] FAILURE components/nasm.bst: Could not find file at /home/fedora/.cache/buildstream/artifacts/extract/freedesktop-sdk/components-asciidoc/35c899591f013697752eea1aa856cac2aa688073f4f129fc7a7a56f44aa3790d/meta/public.yaml
    ======================================================================

[--:--:--][8886a381][ main:components/itstool.bst        ] STATUS  Build suspending
[--:--:--][01c1279d][ main:components/kmod.bst           ] STATUS  Build suspending
[--:--:--][f5db1b4f][ main:components/json-c.bst         ] STATUS  Build suspending

**Comment by juergbi** [comment]
In BuildStream master this should no longer be an issue. buildbox-casd will print a warning if the equivalent cache size file is invalid but will fall back to the slow path of recalculating the cache size (and write a valid file on proper shutdown). Also, the concept of `extract` directories no longer exists in master, so the second issue should not occur anymore either.

As a workaround I'd suggest deleting `~/.cache/buildstream/artifacts/extract`. BuildStream will recreate extract directories as needed.

**Comment by devcurmudgeon** [comment]
thanks for the comment @juergbi - why does there need to be a cache size file at all? this seems a bit fragile to me

**Comment by juergbi** [comment]
It's an optimization of the startup time. For cache expiry to work, BuildStream (or buildbox-casd for master) needs to know the disk usage of the cached blobs. Determining the disk usage can take quite a bit of time if a lot of blobs are cached.

buildbox-casd deletes the cache size file on first write (i.e. when it's invalidated), so a crash of buildbox-casd shouldn't cause any problems except that next startup will take longer. And the cache size file is created using atomic rename, so it should never be corrupt, assuming a reasonable filesystem and storage device.

The cached size will be incorrect if anyone were to manually add or remove blobs without going through buildbox-casd. Is this the case you're worried about?

See also https://gitlab.com/BuildGrid/buildbox/buildbox-casd/-/issues/59

**Comment by devcurmudgeon** [comment]
Thanks for the detailed explanation. I hadn't thought about the manual case you mentioned, let's go with that... what would happen if the stored size is wrong?

Presumably even on the slow path, bst doesn't stop doing other things while the calculation happens?

**Comment by nanonyme** [comment]
I just deleted entire file and everything worked. It looked like it's safe to just force re-computation if we don't have size available. This can typically happen because of eg system crash.

**Comment by gtristan** [comment]
This should be sufficiently fixed by merging #1575, which will fallback to recalculating the cache size on startup in the case of such a parse failure.
